File: backend-spring/src/main/java/com/ecommerce/backend/application/service/OrderService.java
package com.ecommerce.backend.application.service;

import com.ecommerce.backend.application.dto.*;
import com.ecommerce.backend.domain.entity.*;
import com.ecommerce.backend.infrastructure.repository.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@Transactional
public class OrderService {

    @Autowired
    private OrderRepository orderRepository;

    @Autowired
    private OrderItemRepository orderItemRepository;

    @Autowired
    private ProductRepository productRepository;

    @Autowired
    private UserRepository userRepository;

    public BaseResponseDto<List<OrderDto>> getUserOrders(Long userId) {
        try {
            List<Order> orders = orderRepository.findByUserIdAndIsActiveTrueOrderByCreatedAtDesc(userId);
            List<OrderDto> orderDtos = orders.stream()
                    .map(this::convertToOrderDto)
                    .collect(Collectors.toList());

            return BaseResponseDto.success(orderDtos, "Orders retrieved successfully");
        } catch (Exception ex) {
            return BaseResponseDto.error("Error retrieving orders: " + ex.getMessage());
        }
    }

    public BaseResponseDto<OrderDto> getOrderById(Long orderId, Long userId) {
        try {
            Order order = orderRepository.findByIdAndUserIdAndIsActiveTrue(orderId, userId)
                    .orElse(null);

            if (order == null) {
                return BaseResponseDto.error("Order not found");
            }

            OrderDto orderDto = convertToOrderDto(order);
            return BaseResponseDto.success(orderDto, "Order retrieved successfully");
        } catch (Exception ex) {
            return BaseResponseDto.error("Error retrieving order: " + ex.getMessage());
        }
    }

    public BaseResponseDto<OrderDto> createOrder(Long userId, CreateOrderDto createOrderDto) {
        try {
            // Validate products exist and are in stock
            List<Long> productIds = createOrderDto.getOrderItems().stream()
                    .map(CreateOrderItemDto::getProductId)
                    .collect(Collectors.toList());

            List<Product> products = productRepository.findAllById(productIds);
            if (products.size() != productIds.size()) {
                return BaseResponseDto.error("One or more products not found");
            }

            // Check stock availability
            for (CreateOrderItemDto orderItemDto : createOrderDto.getOrderItems()) {
                Product product = products.stream()
                        .filter(p -> p.getId().equals(orderItemDto.getProductId()))
                        .findFirst()
                        .orElse(null);

                if (product == null || product.getUnitInStock() < orderItemDto.getQuantity()) {
                    return BaseResponseDto.error("Insufficient stock for product: " + 
                            (product != null ? product.getProductName() : "Unknown"));
                }
            }

            // Generate order number
            String orderNumber = "ORD-" + LocalDateTime.now().toString().substring(0, 10).replace("-", "") + 
                    "-" + UUID.randomUUID().toString().substring(0, 8).toUpperCase();

            // Calculate total amount
            BigDecimal totalAmount = BigDecimal.ZERO;
            Order order = new Order();
            order.setOrderNumber(orderNumber);
            order.setStatus("Pending");
            order.setShippingAddress(createOrderDto.getShippingAddress());
            order.setBillingAddress(createOrderDto.getBillingAddress() != null ? 
                    createOrderDto.getBillingAddress() : createOrderDto.getShippingAddress());
            order.setUserId(userId);
            order.setIsActive(true);
            order.setCreatedAt(LocalDateTime.now());
            order.setUpdatedAt(LocalDateTime.now());

            // Create order items and calculate total
            for (CreateOrderItemDto orderItemDto : createOrderDto.getOrderItems()) {
                Product product = products.stream()
                        .filter(p -> p.getId().equals(orderItemDto.getProductId()))
                        .findFirst()
                        .orElse(null);

                if (product != null) {
                    BigDecimal discountedPrice = product.getUnitPrice().multiply(
                            BigDecimal.ONE.subtract(product.getDiscount().divide(BigDecimal.valueOf(100))));
                    BigDecimal itemTotal = discountedPrice.multiply(BigDecimal.valueOf(orderItemDto.getQuantity()));
                    totalAmount = totalAmount.add(itemTotal);

                    OrderItem orderItem = new OrderItem();
                    orderItem.setProductId(product.getId());
                    orderItem.setQuantity(orderItemDto.getQuantity());
                    orderItem.setUnitPrice(product.getUnitPrice());
                    orderItem.setDiscount(product.getDiscount());
                    orderItem.setIsActive(true);
                    orderItem.setCreatedAt(LocalDateTime.now());
                    orderItem.setUpdatedAt(LocalDateTime.now());
                    orderItem.setOrder(order);

                    order.getOrderItems().add(orderItem);

                    // Update product stock
                    product.setUnitInStock(product.getUnitInStock() - orderItemDto.getQuantity());
                    product.setUpdatedAt(LocalDateTime.now());
                }
            }

            order.setTotalAmount(totalAmount);
            Order savedOrder = orderRepository.save(order);

            OrderDto orderDto = convertToOrderDto(savedOrder);
            return BaseResponseDto.success(orderDto, "Order created successfully");
        } catch (Exception ex) {
            return BaseResponseDto.error("Error creating order: " + ex.getMessage());
        }
    }

    public BaseResponseDto<OrderDto> updateOrderStatus(Long orderId, Long userId, UpdateOrderStatusDto updateOrderStatusDto) {
        try {
            Order order = orderRepository.findByIdAndUserIdAndIsActiveTrue(orderId, userId)
                    .orElse(null);

            if (order == null) {
                return BaseResponseDto.error("Order not found");
            }

            order.setStatus(updateOrderStatusDto.getStatus());
            order.setUpdatedAt(LocalDateTime.now());
            Order savedOrder = orderRepository.save(order);

            OrderDto orderDto = convertToOrderDto(savedOrder);
            return BaseResponseDto.success(orderDto, "Order status updated successfully");
        } catch (Exception ex) {
            return BaseResponseDto.error("Error updating order status: " + ex.getMessage());
        }
    }

    public BaseResponseDto<String> cancelOrder(Long orderId, Long userId) {
        try {
            Order order = orderRepository.findByIdAndUserIdAndIsActiveTrue(orderId, userId)
                    .orElse(null);

            if (order == null) {
                return BaseResponseDto.error("Order not found");
            }

            if ("Shipped".equals(order.getStatus()) || "Delivered".equals(order.getStatus())) {
                return BaseResponseDto.error("Cannot cancel shipped or delivered orders");
            }

            // Restore product stock
            for (OrderItem orderItem : order.getOrderItems()) {
                Product product = productRepository.findById(orderItem.getProductId()).orElse(null);
                if (product != null) {
                    product.setUnitInStock(product.getUnitInStock() + orderItem.getQuantity());
                    product.setUpdatedAt(LocalDateTime.now());
                }
            }

            order.setStatus("Cancelled");
            order.setUpdatedAt(LocalDateTime.now());
            orderRepository.save(order);

            return BaseResponseDto.success("Order cancelled successfully", "Order cancelled successfully");
        } catch (Exception ex) {
            return BaseResponseDto.error("Error cancelling order: " + ex.getMessage());
        }
    }

    public BaseResponseDto<List<OrderDto>> getAllOrders(int pageNumber, int pageSize) {
        try {
            Pageable pageable = PageRequest.of(pageNumber - 1, pageSize);
            Page<Order> orderPage = orderRepository.findByIsActiveTrueOrderByCreatedAtDesc(pageable);
            
            List<OrderDto> orderDtos = orderPage.getContent().stream()
                    .map(this::convertToOrderDto)
                    .collect(Collectors.toList());

            return BaseResponseDto.success(orderDtos, "Orders retrieved successfully");
        } catch (Exception ex) {
            return BaseResponseDto.error("Error retrieving orders: " + ex.getMessage());
        }
    }

    private OrderDto convertToOrderDto(Order order) {
        List<OrderItemDto> orderItemDtos = order.getOrderItems().stream()
                .map(this::convertToOrderItemDto)
                .collect(Collectors.toList());

        User user = userRepository.findById(order.getUserId()).orElse(null);
        String userName = user != null ? user.getFirstName() + " " + user.getLastName() : "Unknown User";

        return new OrderDto(
                order.getId(),
                order.getOrderNumber(),
                order.getTotalAmount(),
                order.getStatus(),
                order.getShippingAddress(),
                order.getBillingAddress(),
                order.getCreatedAt(),
                order.getUpdatedAt(),
                order.getUserId(),
                userName,
                orderItemDtos
        );
    }

    private OrderItemDto convertToOrderItemDto(OrderItem orderItem) {
        Product product = productRepository.findById(orderItem.getProductId()).orElse(null);
        String productName = product != null ? product.getProductName() : "Unknown Product";
        String productImageUrl = product != null ? product.getImageUrl() : null;

        BigDecimal totalPrice = orderItem.getUnitPrice().multiply(BigDecimal.valueOf(orderItem.getQuantity()))
                .multiply(BigDecimal.ONE.subtract(orderItem.getDiscount().divide(BigDecimal.valueOf(100))));

        return new OrderItemDto(
                orderItem.getId(),
                orderItem.getQuantity(),
                orderItem.getUnitPrice(),
                orderItem.getDiscount(),
                totalPrice,
                orderItem.getProductId(),
                productName,
                productImageUrl,
                orderItem.getOrder().getId()
        );
    }
}

File: backend-dotnet/Application/Services/OrderService.cs
using EcommerceBackend.Application.DTOs;
using EcommerceBackend.Domain.Entities;
using EcommerceBackend.Infrastructure.Data;
using Microsoft.EntityFrameworkCore;

namespace EcommerceBackend.Application.Services
{
    public class OrderService : IOrderService
    {
        private readonly ApplicationDbContext _context;

        public OrderService(ApplicationDbContext context)
        {
            _context = context;
        }

        public async Task<BaseResponseDto<List<OrderDto>>> GetUserOrdersAsync(int userId)
        {
            try
            {
                var orders = await _context.Orders
                    .Where(o => o.UserId == userId && o.IsActive)
                    .Include(o => o.OrderItems)
                        .ThenInclude(oi => oi.Product)
                    .OrderByDescending(o => o.CreatedAt)
                    .Select(o => new OrderDto
                    {
                        Id = o.Id,
                        OrderNumber = o.OrderNumber,
                        TotalAmount = o.TotalAmount,
                        Status = o.Status,
                        ShippingAddress = o.ShippingAddress,
                        BillingAddress = o.BillingAddress,
                        CreatedAt = o.CreatedAt,
                        UpdatedAt = o.UpdatedAt,
                        UserId = o.UserId,
                        UserName = o.User.FirstName + " " + o.User.LastName,
                        OrderItems = o.OrderItems.Select(oi => new OrderItemDto
                        {
                            Id = oi.Id,
                            Quantity = oi.Quantity,
                            UnitPrice = oi.UnitPrice,
                            Discount = oi.Discount,
                            TotalPrice = oi.Quantity * oi.UnitPrice * (1 - oi.Discount / 100),
                            ProductId = oi.ProductId,
                            ProductName = oi.Product.ProductName,
                            ProductImageUrl = oi.Product.ImageUrl,
                            OrderId = oi.OrderId
                        }).ToList()
                    })
                    .ToListAsync();

                return BaseResponseDto<List<OrderDto>>.SuccessResult("Orders retrieved successfully", orders);
            }
            catch (Exception ex)
            {
                return BaseResponseDto<List<OrderDto>>.ErrorResult($"Error retrieving orders: {ex.Message}");
            }
        }

        public async Task<BaseResponseDto<OrderDto>> GetOrderByIdAsync(int orderId, int userId)
        {
            try
            {
                var order = await _context.Orders
                    .Where(o => o.Id == orderId && o.UserId == userId && o.IsActive)
                    .Include(o => o.OrderItems)
                        .ThenInclude(oi => oi.Product)
                    .Select(o => new OrderDto
                    {
                        Id = o.Id,
                        OrderNumber = o.OrderNumber,
                        TotalAmount = o.TotalAmount,
                        Status = o.Status,
                        ShippingAddress = o.ShippingAddress,
                        BillingAddress = o.BillingAddress,
                        CreatedAt = o.CreatedAt,
                        UpdatedAt = o.UpdatedAt,
                        UserId = o.UserId,
                        UserName = o.User.FirstName + " " + o.User.LastName,
                        OrderItems = o.OrderItems.Select(oi => new OrderItemDto
                        {
                            Id = oi.Id,
                            Quantity = oi.Quantity,
                            UnitPrice = oi.UnitPrice,
                            Discount = oi.Discount,
                            TotalPrice = oi.Quantity * oi.UnitPrice * (1 - oi.Discount / 100),
                            ProductId = oi.ProductId,
                            ProductName = oi.Product.ProductName,
                            ProductImageUrl = oi.Product.ImageUrl,
                            OrderId = oi.OrderId
                        }).ToList()
                    })
                    .FirstOrDefaultAsync();

                if (order == null)
                {
                    return BaseResponseDto<OrderDto>.ErrorResult("Order not found");
                }

                return BaseResponseDto<OrderDto>.SuccessResult("Order retrieved successfully", order);
            }
            catch (Exception ex)
            {
                return BaseResponseDto<OrderDto>.ErrorResult($"Error retrieving order: {ex.Message}");
            }
        }

        public async Task<BaseResponseDto<OrderDto>> CreateOrderAsync(int userId, CreateOrderDto createOrderDto)
        {
            try
            {
                // Validate products exist and are in stock
                var productIds = createOrderDto.OrderItems.Select(oi => oi.ProductId).ToList();
                var products = await _context.Products
                    .Where(p => productIds.Contains(p.Id) && p.IsActive)
                    .ToListAsync();

                if (products.Count != productIds.Count)
                {
                    return BaseResponseDto<OrderDto>.ErrorResult("One or more products not found");
                }

                // Check stock availability
                foreach (var orderItem in createOrderDto.OrderItems)
                {
                    var product = products.First(p => p.Id == orderItem.ProductId);
                    if (product.UnitInStock < orderItem.Quantity)
                    {
                        return BaseResponseDto<OrderDto>.ErrorResult($"Insufficient stock for product: {product.ProductName}");
                    }
                }

                // Generate order number
                var orderNumber = $"ORD-{DateTime.UtcNow:yyyyMMdd}-{Guid.NewGuid().ToString("N")[..8].ToUpper()}";

                // Calculate total amount
                decimal totalAmount = 0;
                var orderItems = new List<OrderItem>();

                foreach (var orderItemDto in createOrderDto.OrderItems)
                {
                    var product = products.First(p => p.Id == orderItemDto.ProductId);
                    var discountedPrice = product.UnitPrice * (1 - product.Discount / 100);
                    var itemTotal = orderItemDto.Quantity * discountedPrice;
                    totalAmount += itemTotal;

                    orderItems.Add(new OrderItem
                    {
                        ProductId = orderItemDto.ProductId,
                        Quantity = orderItemDto.Quantity,
                        UnitPrice = product.UnitPrice,
                        Discount = product.Discount,
                        IsActive = true,
                        CreatedAt = DateTime.UtcNow,
                        UpdatedAt = DateTime.UtcNow
                    });

                    // Update product stock
                    product.UnitInStock -= orderItemDto.Quantity;
                    product.UpdatedAt = DateTime.UtcNow;
                }

                // Create order
                var order = new Order
                {
                    OrderNumber = orderNumber,
                    TotalAmount = totalAmount,
                    Status = "Pending",
                    ShippingAddress = createOrderDto.ShippingAddress,
                    BillingAddress = createOrderDto.BillingAddress ?? createOrderDto.ShippingAddress,
                    UserId = userId,
                    IsActive = true,
                    CreatedAt = DateTime.UtcNow,
                    UpdatedAt = DateTime.UtcNow,
                    OrderItems = orderItems
                };

                _context.Orders.Add(order);
                await _context.SaveChangesAsync();

                // Return created order
                var createdOrder = await GetOrderByIdAsync(order.Id, userId);
                return createdOrder;
            }
            catch (Exception ex)
            {
                return BaseResponseDto<OrderDto>.ErrorResult($"Error creating order: {ex.Message}");
            }
        }

        public async Task<BaseResponseDto<OrderDto>> UpdateOrderStatusAsync(int orderId, int userId, UpdateOrderStatusDto updateOrderStatusDto)
        {
            try
            {
                var order = await _context.Orders
                    .Where(o => o.Id == orderId && o.UserId == userId && o.IsActive)
                    .FirstOrDefaultAsync();

                if (order == null)
                {
                    return BaseResponseDto<OrderDto>.ErrorResult("Order not found");
                }

                order.Status = updateOrderStatusDto.Status;
                order.UpdatedAt = DateTime.UtcNow;

                await _context.SaveChangesAsync();

                return await GetOrderByIdAsync(orderId, userId);
            }
            catch (Exception ex)
            {
                return BaseResponseDto<OrderDto>.ErrorResult($"Error updating order status: {ex.Message}");
            }
        }

        public async Task<BaseResponseDto<string>> CancelOrderAsync(int orderId, int userId)
        {
            try
            {
                var order = await _context.Orders
                    .Where(o => o.Id == orderId && o.UserId == userId && o.IsActive)
                    .Include(o => o.OrderItems)
                        .ThenInclude(oi => oi.Product)
                    .FirstOrDefaultAsync();

                if (order == null)
                {
                    return BaseResponseDto<string>.ErrorResult("Order not found");
                }

                if (order.Status == "Shipped" || order.Status == "Delivered")
                {
                    return BaseResponseDto<string>.ErrorResult("Cannot cancel shipped or delivered orders");
                }

                // Restore product stock
                foreach (var orderItem in order.OrderItems)
                {
                    var product = await _context.Products.FindAsync(orderItem.ProductId);
                    if (product != null)
                    {
                        product.UnitInStock += orderItem.Quantity;
                        product.UpdatedAt = DateTime.UtcNow;
                    }
                }

                order.Status = "Cancelled";
                order.UpdatedAt = DateTime.UtcNow;

                await _context.SaveChangesAsync();

                return BaseResponseDto<string>.SuccessResult("Order cancelled successfully", "Order cancelled successfully");
            }
            catch (Exception ex)
            {
                return BaseResponseDto<string>.ErrorResult($"Error cancelling order: {ex.Message}");
            }
        }

        public async Task<BaseResponseDto<List<OrderDto>>> GetAllOrdersAsync(int pageNumber = 1, int pageSize = 10)
        {
            try
            {
                var orders = await _context.Orders
                    .Where(o => o.IsActive)
                    .Include(o => o.OrderItems)
                        .ThenInclude(oi => oi.Product)
                    .Include(o => o.User)
                    .OrderByDescending(o => o.CreatedAt)
                    .Skip((pageNumber - 1) * pageSize)
                    .Take(pageSize)
                    .Select(o => new OrderDto
                    {
                        Id = o.Id,
                        OrderNumber = o.OrderNumber,
                        TotalAmount = o.TotalAmount,
                        Status = o.Status,
                        ShippingAddress = o.ShippingAddress,
                        BillingAddress = o.BillingAddress,
                        CreatedAt = o.CreatedAt,
                        UpdatedAt = o.UpdatedAt,
                        UserId = o.UserId,
                        UserName = o.User.FirstName + " " + o.User.LastName,
                        OrderItems = o.OrderItems.Select(oi => new OrderItemDto
                        {
                            Id = oi.Id,
                            Quantity = oi.Quantity,
                            UnitPrice = oi.UnitPrice,
                            Discount = oi.Discount,
                            TotalPrice = oi.Quantity * oi.UnitPrice * (1 - oi.Discount / 100),
                            ProductId = oi.ProductId,
                            ProductName = oi.Product.ProductName,
                            ProductImageUrl = oi.Product.ImageUrl,
                            OrderId = oi.OrderId
                        }).ToList()
                    })
                    .ToListAsync();

                return BaseResponseDto<List<OrderDto>>.SuccessResult("Orders retrieved successfully", orders);
            }
            catch (Exception ex)
            {
                return BaseResponseDto<List<OrderDto>>.ErrorResult($"Error retrieving orders: {ex.Message}");
            }
        }
    }
}
